utils/log_parser.py

```diff
@@ -0,0 +1,311 @@
+# Copyright (c) 2017-present, Facebook, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import re
+
+# Maintaining all nodes
+NODES_MAP = {}
+
+# Maintaining all edges
+EDGES_MAP = {}
+
+# Scope stack
+SCOPE_STACK = []
+
+# Critical Phases
+CRITICAL_PHASES = ["lower", "optimize", "fold", "optimizeFunction",
+                   "optimizeFunctionBeforeLowering", "transformPostLowering"]
+
+
+class Node:
+    def __init__(self, kindName, name):
+        self.kindName_ = kindName
+        self.name_ = name
+        self.inputs_ = []
+        self.users_ = {}
+
+    def get_kind_name(self):
+        return self.kindName_
+
+    def get_name(self):
+        return self.name_
+
+    def get_inputs(self):
+        return self.inputs_
+
+    def get_users(self):
+        return self.users_
+
+    def add_user(self, u):
+        if u not in self.users_:
```

Comment:
You could simplify this using `DefaultDict`: https://docs.python.org/2/library/collections.html#collections.defaultdict